File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/PhysicalDamageTypeEnum.java
package ai.ragu.charactermanager.enumeration;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.HashMap;
import java.util.Map;

public enum PhysicalDamageTypeEnum implements DamageTypeEnum {
    BLUDGEONING("bludgeoning"),
    PIERCING("piercing"),
    SLASHING("slashing");

    private final String value;
    private final static Map<String, PhysicalDamageTypeEnum> CONSTANTS = new HashMap<>();

    static {
        for (PhysicalDamageTypeEnum c: values()) {
            CONSTANTS.put(c.value, c);
        }
    }

    PhysicalDamageTypeEnum(String value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return this.value;
    }

    @JsonValue
    public String value() {
        return this.value;
    }

    @JsonCreator
    public static PhysicalDamageTypeEnum fromValue(String value) {
        PhysicalDamageTypeEnum constant = CONSTANTS.get(value);
        if (constant == null) {
            throw new IllegalArgumentException(value);
        } else {
            return constant;
        }
    }
}

File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/PhysicalDamageTypeEnum.kt
package ai.ragu.charactermanager.enumeration

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import java.util.*

enum class PhysicalDamageTypeEnum(private val value: String) : DamageTypeEnum {
    BLUDGEONING("bludgeoning"),
    PIERCING("piercing"),
    SLASHING("slashing");

    companion object {
        private val CONSTANTS: MutableMap<String, PhysicalDamageTypeEnum> = HashMap()

        @JsonCreator
        fun fromValue(value: String?): PhysicalDamageTypeEnum {
            val constant = CONSTANTS[value]
            return constant ?: throw IllegalArgumentException(value)
        }

        init {
            for (c in values()) {
                CONSTANTS[c.value] = c
            }
        }
    }

    override fun toString(): String {
        return value
    }

    @JsonValue
    fun value(): String {
        return value
    }

}